Add infiniband monitoring to collector proc.plugin

Saruspete opened the pull request:
##### Summary
Implement infiniband monitoring. Fixes #2620 

Infiniband is a fabric heavily used in HPC and supercomputers. It's designed for short range (local to  datacenter), provides RDMA, low latency communication and is handled directly by MPI libraries.

It often is coupled with "IPoIB" (IP over InfiniBand) which adds the IP connectivity over Infiniband messaging and is also seen as a simple network card. However, the MPI libraries doesn't use it, and talk directly to IB cards. 

From the drivers, there's no single file that provides details and summary over all IB interfaces, so we have to parse `/sys/class/infiniband/<verb>/ports/<portid>/counters/` path to get the counters provided by driver.
Each statistics are per-port.

This prevents to use the `procfile` helpers. As there is a lot of single statistics files (21 for `counters`, and more than 120 in `hw_counters` for OmniPath), I used macros to generate the repetitive elements over `read_single_number_file`

The `counters` are grouped in 3 categories (graphs): `bytes` `packets` and `errors`.

The `hw_counters` needs to be done if possible, as it seems very interface-specific. More feedback will be needed to have more output.

##### Component Name
collectors/proc.plugin

##### Test Plan

Emulates infiniband hw:

Change default path with netdata.conf to use a fake folder:
```
[plugin:proc:/sys/class/infiniband]
  dirname to monitor = /opt/netdata/sys/class/infiniband
```

Create a fake tree and populate it
```bash
typeset ibpath="/opt/netdata/sys/class/infiniband/test1_1/ports/99/counters/"
mkdir -p "$ibpath"
typeset -A counters
# Generate fake traffic
while sleep 1; do
  for counter in {port,multicast,unicast}_{rcv,xmit}_packets port_{rcv,xmit}_data port_{rcv_{,constraint_,remote_physical_,switch_relay_}errors,xmit_{discards,wait,constraint_errors}} link_{downed,error_recovery}  local_link_integrity_errors excessive_buffer_overrun_errors VL15_dropped symbol_error; do
    counters[$counter]=$((${counters[$counter]} + $RANDOM % 100))
    echo ${counters[$counter]} > "$ibpath/$counter"
  done
done
```

Create a state file
``` bash
echo "4: ACTIVE" > /opt/netdata/sys/class/infiniband/test1_1/state
```

That must create 3 graphs, under family "infiniband" / "test1_1-99"

##### Additional Information

ilyam8 commented:
> The hw_counters needs to be done if possible, as it seems very interface-specific. More feedback will be needed to have more output.

@mtds i see [you are interested](https://github.com/openhpc/ohpc/issues/1159#issuecomment-640045110) in this collector, any chance you can help @Saruspete to finish it?

mtds commented:
> > The hw_counters needs to be done if possible, as it seems very interface-specific. More feedback will be needed to have more output.
> 
> @mtds i see [you are interested](https://github.com/openhpc/ohpc/issues/1159#issuecomment-640045110) in this collector, any chance you can help @Saruspete to finish it?

@ilyam8: I would love to help but in this period I am pretty tight with my time, so I cannot promise you anything but I will nevertheless try this additional plugin and give back some feedback. I have access to different IB cards from Mellanox but we do not have Omnipath HW

ilyam8 commented:
>  I have access to different IB cards from Mellanox

Sounds great, then please check this PR and provide some feedback when you have time 👍

Saruspete commented:
I now have access to Mellanox Connect-X 5 Ex and Connect-X 6 cards.
Seems it's mostly the omnipath cards that provides extensive statistics, most of which are off by default, so of little interest:

<details>
  <summary>Omnipath</summary>
<pre>
cd /sys/class/infiniband/hfi1_0/ports/1/hw_counters
for i in *; do printf "%-20.20s  %s\n" "$i" "$(<$i)"; done
DmaWait               50646
lifespan              10
LoopPkts              115262
OtherNak              0
PktDrop               134
RcAcks                19918731
RcDelayComp           3109215
RcDupRew              0
RcQacks               233
RcResend              2079
RcSeqNak              0
RcTimeOut             0
RcvCstrErr            0
RcvHdrOvr0            0
RcvHdrOvr1            0
RcvHdrOvr10           0
RcvHdrOvr100          0
RcvHdrOvr101          0
RcvHdrOvr102          0
RcvHdrOvr103          0
RcvHdrOvr104          0
RcvHdrOvr105          0
RcvHdrOvr106          0
RcvHdrOvr11           0
RcvHdrOvr12           0
RcvHdrOvr13           0
RcvHdrOvr14           0
RcvHdrOvr15           0
RcvHdrOvr16           0
RcvHdrOvr17           0
RcvHdrOvr18           0
RcvHdrOvr19           0
RcvHdrOvr2            0
RcvHdrOvr20           0
RcvHdrOvr21           0
RcvHdrOvr22           0
RcvHdrOvr23           0
RcvHdrOvr24           0
RcvHdrOvr25           0
RcvHdrOvr26           0
RcvHdrOvr27           0
RcvHdrOvr28           0
RcvHdrOvr29           0
RcvHdrOvr3            0
RcvHdrOvr30           0
RcvHdrOvr31           0
RcvHdrOvr32           0
RcvHdrOvr33           0
RcvHdrOvr34           0
RcvHdrOvr35           0
RcvHdrOvr36           0
RcvHdrOvr37           0
RcvHdrOvr38           0
RcvHdrOvr39           0
RcvHdrOvr4            0
RcvHdrOvr40           0
RcvHdrOvr41           0
RcvHdrOvr42           0
RcvHdrOvr43           0
RcvHdrOvr44           0
RcvHdrOvr45           0
RcvHdrOvr46           0
RcvHdrOvr47           0
RcvHdrOvr48           0
RcvHdrOvr49           0
RcvHdrOvr5            0
RcvHdrOvr50           0
RcvHdrOvr51           0
RcvHdrOvr52           0
RcvHdrOvr53           0
RcvHdrOvr54           0
RcvHdrOvr55           0
RcvHdrOvr56           0
RcvHdrOvr57           0
RcvHdrOvr58           0
RcvHdrOvr59           0
RcvHdrOvr6            0
RcvHdrOvr60           0
RcvHdrOvr61           0
RcvHdrOvr62           0
RcvHdrOvr63           0
RcvHdrOvr64           0
RcvHdrOvr65           0
RcvHdrOvr66           0
RcvHdrOvr67           0
RcvHdrOvr68           0
RcvHdrOvr69           0
RcvHdrOvr7            0
RcvHdrOvr70           0
RcvHdrOvr71           0
RcvHdrOvr72           0
RcvHdrOvr73           0
RcvHdrOvr74           0
RcvHdrOvr75           0
RcvHdrOvr76           0
RcvHdrOvr77           0
RcvHdrOvr78           0
RcvHdrOvr79           0
RcvHdrOvr8            0
RcvHdrOvr80           0
RcvHdrOvr81           0
RcvHdrOvr82           0
RcvHdrOvr83           0
RcvHdrOvr84           0
RcvHdrOvr85           0
RcvHdrOvr86           0
RcvHdrOvr87           0
RcvHdrOvr88           0
RcvHdrOvr89           0
RcvHdrOvr9            0
RcvHdrOvr90           0
RcvHdrOvr91           0
RcvHdrOvr92           0
RcvHdrOvr93           0
RcvHdrOvr94           0
RcvHdrOvr95           0
RcvHdrOvr96           0
RcvHdrOvr97           0
RcvHdrOvr98           0
RcvHdrOvr99           0
RdmaSeq               0
RnrNak                0
RxPkt                 24047676428
RxWords               32265824464869
SeqNak                18
SwLinkDown,32         0
SwLinkUp,32           0
TxDropped,32          0
TxFlitVL0             16460745601437
TxFlitVL1             0
TxFlitVL15            34192640
TxFlitVL2             0
TxFlitVL3             0
TxFlitVL4             0
TxFlitVL5             0
TxFlitVL6             0
TxFlitVL7             0
TxFlowStall,32        514155663
TxHdrErr,32           0
TxInvalLen,32         0
TxMMLenErr,32         0
TxPkt                 24007563078
TxPktVL0              24007098469
TxPktVL1              0
TxPktVL15             464609
TxPktVL2              0
TxPktVL3              0
TxPktVL4              0
TxPktVL5              0
TxPktVL6              0
TxPktVL7              0
TxUnderrun,32         0
TxUnVLErr,32          0
TxWait                2163652785833
TxWaitVL0             2160892821953
TxWaitVL1             0
TxWaitVL15            0
TxWaitVL2             0
TxWaitVL3             0
TxWaitVL4             0
TxWaitVL5             0
TxWaitVL6             0
TxWaitVL7             0
TxWords               32898345374313
Unaligned             0
UnknownFrame          0
XmitCstrErr           0
XmitDscd,32           0
XmitDscdVl0,32        0
XmitDscdVl1,32        0
XmitDscdVl15,32       0
XmitDscdVl2,32        0
XmitDscdVl3,32        0
XmitDscdVl4,32        0
XmitDscdVl5,32        0
XmitDscdVl6,32        0
XmitDscdVl7,32        0

</pre>
</details>

<details>
  <summary>ConnectX-5 Ex CX556A</summary>
<pre>
cd /sys/class/infiniband/mlx5_0/ports/1/hw_counters
for i in *; do printf "%-30.30s  %s\n" "$i" "$(<$i)"; done
duplicate_request               0
implied_nak_seq_err             0
lifespan                        10
local_ack_timeout_err           0
np_cnp_sent                     367283
np_ecn_marked_roce_packets      369554
out_of_buffer                   0
out_of_sequence                 0
packet_seq_err                  0
req_cqe_error                   0
req_cqe_flush_error             0
req_remote_access_errors        0
req_remote_invalid_request      0
resp_cqe_error                  606208
resp_cqe_flush_error            606060
resp_local_length_error         0
resp_remote_access_errors       0
rnr_nak_retry_err               0
rp_cnp_handled                  242451
rp_cnp_ignored                  0
rx_atomic_requests              0
rx_dct_connect                  0
rx_icrc_encapsulated            0
rx_read_requests                41428
rx_write_requests               0
</pre>
</details>

<details>
  <summary>ConnectX-6 MCX653105A-ECAT</summary>
<pre>
cd /sys/class/infiniband/mlx5_0/ports/1/hw_counters
for i in *; do printf "%-30.30s  %s\n" "$i" "$(<$i)"; done
duplicate_request               0
implied_nak_seq_err             0
lifespan                        10
local_ack_timeout_err           0
out_of_buffer                   0
out_of_sequence                 0
packet_seq_err                  0
req_cqe_error                   0
req_cqe_flush_error             0
req_remote_access_errors        0
req_remote_invalid_request      0
resp_cqe_error                  0
resp_cqe_flush_error            0
resp_local_length_error         0
resp_remote_access_errors       0
rnr_nak_retry_err               0
roce_adp_retrans                0
roce_adp_retrans_to             0
roce_slow_restart               0
roce_slow_restart_cnps          0
roce_slow_restart_trans         0
rx_atomic_requests              0
rx_dct_connect                  0
rx_icrc_encapsulated            0
rx_read_requests                0
rx_write_requests               0
</pre>
</details>

Saruspete commented:
I would like your insights here:

The HWCounters are very vendor-specific, and I can't get a standard way to fetch them. So, I'm only implementing Mellanox (mlx / mlnx, the standard for Infiniband) for the time being, having all setup to add easily new vendors.

I'm trying to do as follow:

The macro will generate:
- one struct `ibporthw_<vendor>`  per vendor. it'll contain pointers and paths to the different files, just like the standard counters.
- one function `infiniband_parse_hwcounter_<vendor>`. It'll parse the different files, vendor specific.
**Added size**: 1 struct definition per vendor + 1 function per vendor
**Wasted computation per run**: none

In the `ibport` struct, declare:
- one pointer to each of these struct (to avoid future cast)
- one function pointer to the vendor-specific parsing function
**Added Size**: struct pointer of vendor specific `ibporthw_<vendor>` struct
**Wasted computation per run**: none

On the main `do_sys_class_infiniband`:
- Test if function pointer is non-null, and call it if so.
- Test if RRD graphs are non-null, and fill them if so
**Added Size**: size of vendor-specific `ibporthw_<vendor>` struct
**Wasted computation per run**: test if function-pointer is null & if rrd-graph is null

That way, using function pointer I get minimal runtime overhead (like a series of useless `strcmp`) and I only allocate memory for struct that are effectively handled.

What do you think ? I really do hope you're no too frightened by the macro generation of the multiple files definition.

Saruspete commented:
Took longer than expected, but looks pretty good for now.

![netdata_collector_infiniband](https://user-images.githubusercontent.com/9157986/84093697-3936ce80-a9fb-11ea-99be-5d133d13a37c.png)

I had to parse the `rate` file, to get the width, as counters are specified per-lane. So to get the total bandwidth, I must multiply the counter by the width.

By adding parsing of only active ports, I removed useless `open()` syscalls. I'm not fond of the number of `open` this generates, but seems it's the only way for `sysfs` to refresh their values, unlike `procfs` (check with `sudo strace -fyp "$(pgrep netdata)" -e openat`).

<details>
<summary>sudo strace -fyp "$(pgrep netdata)" -e open</summary>
<pre>
[pid 34187] open("/sys/class/infiniband/mlx5_0/ports/1/counters/port_rcv_data", O_RDONLY) = 76</sys/devices/pci0000:5d/0000:5d:00.0/0000:5e:00.0/infiniband/mlx5_0/ports/1/counters/port_rcv_data>
[pid 34187] open("/sys/class/infiniband/mlx5_0/ports/1/counters/port_xmit_data", O_RDONLY) = 76</sys/devices/pci0000:5d/0000:5d:00.0/0000:5e:00.0/infiniband/mlx5_0/ports/1/counters/port_xmit_data>
[pid 34187] open("/sys/class/infiniband/mlx5_0/ports/1/counters/port_rcv_packets", O_RDONLY) = 76</sys/devices/pci0000:5d/0000:5d:00.0/0000:5e:00.0/infiniband/mlx5_0/ports/1/counters/port_rcv_packets>
[pid 34187] open("/sys/class/infiniband/mlx5_0/ports/1/counters/port_xmit_packets", O_RDONLY) = 76</sys/devices/pci0000:5d/0000:5d:00.0/0000:5e:00.0/infiniband/mlx5_0/ports/1/counters/port_xmit_packets>
[pid 34187] open("/sys/class/infiniband/mlx5_0/ports/1/counters/multicast_rcv_packets", O_RDONLY) = 76</sys/devices/pci0000:5d/0000:5d:00.0/0000:5e:00.0/infiniband/mlx5_0/ports/1/counters/multicast_rcv_packets>
[pid 34187] open("/sys/class/infiniband/mlx5_0/ports/1/counters/multicast_xmit_packets", O_RDONLY) = 76</sys/devices/pci0000:5d/0000:5d:00.0/0000:5e:00.0/infiniband/mlx5_0/ports/1/counters/multicast_xmit_packets>
[pid 34187] open("/sys/class/infiniband/mlx5_0/ports/1/counters/unicast_rcv_packets", O_RDONLY) = 76</sys/devices/pci0000:5d/0000:5d:00.0/0000:5e:00.0/infiniband/mlx5_0/ports/1/counters/unicast_rcv_packets>
[pid 34187] open("/sys/class/infiniband/mlx5_0/ports/1/counters/unicast_xmit_packets", O_RDONLY) = 76</sys/devices/pci0000:5d/0000:5d:00.0/0000:5e:00.0/infiniband/mlx5_0/ports/1/counters/unicast_xmit_packets>
[pid 34187] open("/sys/class/infiniband/mlx5_0/ports/1/counters/port_rcv_errors", O_RDONLY) = 76</sys/devices/pci0000:5d/0000:5d:00.0/0000:5e:00.0/infiniband/mlx5_0/ports/1/counters/port_rcv_errors>
[pid 34187] open("/sys/class/infiniband/mlx5_0/ports/1/counters/port_rcv_constraint_errors", O_RDONLY) = 76</sys/devices/pci0000:5d/0000:5d:00.0/0000:5e:00.0/infiniband/mlx5_0/ports/1/counters/port_rcv_constraint_errors>
[pid 34187] open("/sys/class/infiniband/mlx5_0/ports/1/counters/port_xmit_discards", O_RDONLY) = 76</sys/devices/pci0000:5d/0000:5d:00.0/0000:5e:00.0/infiniband/mlx5_0/ports/1/counters/port_xmit_discards>
[pid 34187] open("/sys/class/infiniband/mlx5_0/ports/1/counters/port_xmit_wait", O_RDONLY) = 76</sys/devices/pci0000:5d/0000:5d:00.0/0000:5e:00.0/infiniband/mlx5_0/ports/1/counters/port_xmit_wait>
[pid 34187] open("/sys/class/infiniband/mlx5_0/ports/1/counters/VL15_dropped", O_RDONLY) = 76</sys/devices/pci0000:5d/0000:5d:00.0/0000:5e:00.0/infiniband/mlx5_0/ports/1/counters/VL15_dropped>
[pid 34187] open("/sys/class/infiniband/mlx5_0/ports/1/counters/excessive_buffer_overrun_errors", O_RDONLY) = 76</sys/devices/pci0000:5d/0000:5d:00.0/0000:5e:00.0/infiniband/mlx5_0/ports/1/counters/excessive_buffer_overrun_errors>
[pid 34187] open("/sys/class/infiniband/mlx5_0/ports/1/counters/link_downed", O_RDONLY) = 76</sys/devices/pci0000:5d/0000:5d:00.0/0000:5e:00.0/infiniband/mlx5_0/ports/1/counters/link_downed>
[pid 34187] open("/sys/class/infiniband/mlx5_0/ports/1/counters/link_error_recovery", O_RDONLY) = 76</sys/devices/pci0000:5d/0000:5d:00.0/0000:5e:00.0/infiniband/mlx5_0/ports/1/counters/link_error_recovery>
[pid 34187] open("/sys/class/infiniband/mlx5_0/ports/1/counters/local_link_integrity_errors", O_RDONLY) = 76</sys/devices/pci0000:5d/0000:5d:00.0/0000:5e:00.0/infiniband/mlx5_0/ports/1/counters/local_link_integrity_errors>
[pid 34187] open("/sys/class/infiniband/mlx5_0/ports/1/counters/symbol_error", O_RDONLY) = 76</sys/devices/pci0000:5d/0000:5d:00.0/0000:5e:00.0/infiniband/mlx5_0/ports/1/counters/symbol_error>
[pid 34187] open("/sys/class/infiniband/mlx5_0/ports/1/counters/port_rcv_remote_physical_errors", O_RDONLY) = 76</sys/devices/pci0000:5d/0000:5d:00.0/0000:5e:00.0/infiniband/mlx5_0/ports/1/counters/port_rcv_remote_physical_errors>
[pid 34187] open("/sys/class/infiniband/mlx5_0/ports/1/counters/port_rcv_switch_relay_errors", O_RDONLY) = 76</sys/devices/pci0000:5d/0000:5d:00.0/0000:5e:00.0/infiniband/mlx5_0/ports/1/counters/port_rcv_switch_relay_errors>
[pid 34187] open("/sys/class/infiniband/mlx5_0/ports/1/counters/port_xmit_constraint_errors", O_RDONLY) = 76</sys/devices/pci0000:5d/0000:5d:00.0/0000:5e:00.0/infiniband/mlx5_0/ports/1/counters/port_xmit_constraint_errors>
[pid 34187] open("/sys/class/infiniband/mlx5_0/ports/1/hw_counters/duplicate_request", O_RDONLY) = 76</sys/devices/pci0000:5d/0000:5d:00.0/0000:5e:00.0/infiniband/mlx5_0/ports/1/hw_counters/duplicate_request>
[pid 34187] open("/sys/class/infiniband/mlx5_0/ports/1/hw_counters/implied_nak_seq_err", O_RDONLY) = 76</sys/devices/pci0000:5d/0000:5d:00.0/0000:5e:00.0/infiniband/mlx5_0/ports/1/hw_counters/implied_nak_seq_err>
[pid 34187] open("/sys/class/infiniband/mlx5_0/ports/1/hw_counters/local_ack_timeout_err", O_RDONLY) = 76</sys/devices/pci0000:5d/0000:5d:00.0/0000:5e:00.0/infiniband/mlx5_0/ports/1/hw_counters/local_ack_timeout_err>
[pid 34187] open("/sys/class/infiniband/mlx5_0/ports/1/hw_counters/out_of_buffer", O_RDONLY) = 76</sys/devices/pci0000:5d/0000:5d:00.0/0000:5e:00.0/infiniband/mlx5_0/ports/1/hw_counters/out_of_buffer>
[pid 34187] open("/sys/class/infiniband/mlx5_0/ports/1/hw_counters/out_of_sequence", O_RDONLY) = 76</sys/devices/pci0000:5d/0000:5d:00.0/0000:5e:00.0/infiniband/mlx5_0/ports/1/hw_counters/out_of_sequence>
[pid 34187] open("/sys/class/infiniband/mlx5_0/ports/1/hw_counters/packet_seq_err", O_RDONLY) = 76</sys/devices/pci0000:5d/0000:5d:00.0/0000:5e:00.0/infiniband/mlx5_0/ports/1/hw_counters/packet_seq_err>
[pid 34187] open("/sys/class/infiniband/mlx5_0/ports/1/hw_counters/req_cqe_error", O_RDONLY) = 76</sys/devices/pci0000:5d/0000:5d:00.0/0000:5e:00.0/infiniband/mlx5_0/ports/1/hw_counters/req_cqe_error>
[pid 34187] open("/sys/class/infiniband/mlx5_0/ports/1/hw_counters/resp_cqe_error", O_RDONLY) = 76</sys/devices/pci0000:5d/0000:5d:00.0/0000:5e:00.0/infiniband/mlx5_0/ports/1/hw_counters/resp_cqe_error>
[pid 34187] open("/sys/class/infiniband/mlx5_0/ports/1/hw_counters/req_cqe_flush_error", O_RDONLY) = 76</sys/devices/pci0000:5d/0000:5d:00.0/0000:5e:00.0/infiniband/mlx5_0/ports/1/hw_counters/req_cqe_flush_error>
[pid 34187] open("/sys/class/infiniband/mlx5_0/ports/1/hw_counters/resp_cqe_flush_error", O_RDONLY) = 76</sys/devices/pci0000:5d/0000:5d:00.0/0000:5e:00.0/infiniband/mlx5_0/ports/1/hw_counters/resp_cqe_flush_error>
[pid 34187] open("/sys/class/infiniband/mlx5_0/ports/1/hw_counters/req_remote_access_errors", O_RDONLY) = 76</sys/devices/pci0000:5d/0000:5d:00.0/0000:5e:00.0/infiniband/mlx5_0/ports/1/hw_counters/req_remote_access_errors>
[pid 34187] open("/sys/class/infiniband/mlx5_0/ports/1/hw_counters/resp_remote_access_errors", O_RDONLY) = 76</sys/devices/pci0000:5d/0000:5d:00.0/0000:5e:00.0/infiniband/mlx5_0/ports/1/hw_counters/resp_remote_access_errors>
[pid 34187] open("/sys/class/infiniband/mlx5_0/ports/1/hw_counters/req_remote_invalid_request", O_RDONLY) = 76</sys/devices/pci0000:5d/0000:5d:00.0/0000:5e:00.0/infiniband/mlx5_0/ports/1/hw_counters/req_remote_invalid_request>
[pid 34187] open("/sys/class/infiniband/mlx5_0/ports/1/hw_counters/resp_local_length_error", O_RDONLY) = 76</sys/devices/pci0000:5d/0000:5d:00.0/0000:5e:00.0/infiniband/mlx5_0/ports/1/hw_counters/resp_local_length_error>
[pid 34187] open("/sys/class/infiniband/mlx5_0/ports/1/hw_counters/rnr_nak_retry_err", O_RDONLY) = 76</sys/devices/pci0000:5d/0000:5d:00.0/0000:5e:00.0/infiniband/mlx5_0/ports/1/hw_counters/rnr_nak_retry_err>
[pid 34187] open("/sys/class/infiniband/mlx5_0/ports/1/hw_counters/rp_cnp_ignored", O_RDONLY) = 76</sys/devices/pci0000:5d/0000:5d:00.0/0000:5e:00.0/infiniband/mlx5_0/ports/1/hw_counters/rp_cnp_ignored>
[pid 34187] open("/sys/class/infiniband/mlx5_0/ports/1/hw_counters/rx_icrc_encapsulated", O_RDONLY) = 76</sys/devices/pci0000:5d/0000:5d:00.0/0000:5e:00.0/infiniband/mlx5_0/ports/1/hw_counters/rx_icrc_encapsulated>
[pid 34187] open("/sys/class/infiniband/mlx5_0/ports/1/hw_counters/np_cnp_sent", O_RDONLY) = 76</sys/devices/pci0000:5d/0000:5d:00.0/0000:5e:00.0/infiniband/mlx5_0/ports/1/hw_counters/np_cnp_sent>
[pid 34187] open("/sys/class/infiniband/mlx5_0/ports/1/hw_counters/np_ecn_marked_roce_packets", O_RDONLY) = 76</sys/devices/pci0000:5d/0000:5d:00.0/0000:5e:00.0/infiniband/mlx5_0/ports/1/hw_counters/np_ecn_marked_roce_packets>
[pid 34187] open("/sys/class/infiniband/mlx5_0/ports/1/hw_counters/rp_cnp_handled", O_RDONLY) = 76</sys/devices/pci0000:5d/0000:5d:00.0/0000:5e:00.0/infiniband/mlx5_0/ports/1/hw_counters/rp_cnp_handled>
[pid 34187] open("/sys/class/infiniband/mlx5_0/ports/1/hw_counters/rx_atomic_requests", O_RDONLY) = 76</sys/devices/pci0000:5d/0000:5d:00.0/0000:5e:00.0/infiniband/mlx5_0/ports/1/hw_counters/rx_atomic_requests>
[pid 34187] open("/sys/class/infiniband/mlx5_0/ports/1/hw_counters/rx_dct_connect", O_RDONLY) = 76</sys/devices/pci0000:5d/0000:5d:00.0/0000:5e:00.0/infiniband/mlx5_0/ports/1/hw_counters/rx_dct_connect>
[pid 34187] open("/sys/class/infiniband/mlx5_0/ports/1/hw_counters/rx_read_requests", O_RDONLY) = 76</sys/devices/pci0000:5d/0000:5d:00.0/0000:5e:00.0/infiniband/mlx5_0/ports/1/hw_counters/rx_read_requests>
[pid 34187] open("/sys/class/infiniband/mlx5_0/ports/1/hw_counters/rx_write_requests", O_RDONLY) = 76</sys/devices/pci0000:5d/0000:5d:00.0/0000:5e:00.0/infiniband/mlx5_0/ports/1/hw_counters/rx_write_requests>
</pre>
</details>

Saruspete commented on collectors/proc.plugin/sys_class_infiniband.c:
```diff
@@ -0,0 +1,721 @@
+				rrdset_flag_set(port->st_bytes, RRDSET_FLAG_DETAIL);
+				// On this chart, we want to have a KB/s so the dashboard will autoscale it
+				// The reported values are also per-lane, so we must multiply it by the width
+				FOREACH_COUNTER_BYTES(GEN_RRD_DIM_ADD_CUSTOM, port, 8 * port->width, 1024, RRD_ALGORITHM_INCREMENTAL)
```
I just have a double about the multiplier `port->width` is working, as I didn't see any higher than 24Gb/s on my test. 
I don't know if I didn't get the MPI benchmark with the correct options to saturate the port, or if the multiplier is not taken into account. From the "info" log line, it's correctly reported as 4.

Saruspete commented:
messed up my git... again... sorry for the noise.

So checked with a double port Omnipath, and with Mellanox. Looks good to me.

ilyam8 commented:
@Saruspete 

1. `Bandwidth usage` chart. I see there is `Link speed` direction? Why we need it? Does it change during runtime?

2. `Packets statistics`. I think `untits` shoude `packets/s`/`pps`. Perhaphs it would be better to have Unicast as a separate chart? Is it even important? Does infiband uses mcast/brcast?

3. 

> I'm not fond of the number of open this generates

How many open files? Does the collector keep them open?

---

Awesome work, you rock!

Saruspete commented:
> 1. `Bandwidth usage` chart. I see there is `Link speed` direction? Why we need it? Does it change during runtime?

It may change, mostly by command `ibportstate` (not aware of any error rate that would downgrade its speed though). So I've added it here to be available on health check, and see what percentage of the max theoretical speed is reached. 
But as having a plain area like this may be confusing, I've also added a configuration item to disable it.

> 2. `Packets statistics`. I think `untits` shoude `packets/s`/`pps`. 

It's the case: `"Packets/s"` line 336. It's one of the units the chart understand and rescale, so during that test, it was in Mpps.

> Perhaphs it would be better to have Unicast as a separate chart? Is it even important? Does infiband uses mcast/brcast?

Yes, there's Ucast & Mcast (lines 26-32), but I've left the "hide 0 dims" checked (my test didn't use mcast)
 
> > I'm not fond of the number of open this generates
> 
> How many open files? Does the collector keep them open?

It only calls `read_single_number_file`, so 1 at a time, but it' 21 `open()` + `close()` for the counters. I don't know if it's high or not for you, but I'm always shy adding syscalls to `proc.plugin`... but here, there's no other way.

> Awesome work, you rock!

It's @ktsaou and all of you in the team that did all the real hard work :)

ilyam8 commented:
> So I've added it here to be available on health check

Could be a variable then. It will do for healthchecks.

Saruspete commented:
Good idea ! never used variables before, but should do the job.

vlvkobal commented:
Please update the proc.plugin documentation.

vlvkobal commented:
We don't use tabs. There should be spaces after commas. Please look at [`.clang-format`](https://github.com/netdata/netdata/blob/master/.clang-format).

Saruspete commented:
Added doc and changed formatting.

In commit `3b01478` I also did a configuration update, but clang-format changed the file, while I was only expecting report, not having the file changed.

vlvkobal commented on collectors/proc.plugin/README.md:
```diff
@@ -462,6 +463,48 @@ and metrics:
+    - Link: downed, recovered, integrity error, minor error
+    - Other events: Tick Wait to send, buffer overrun
+
+If your vendor is supported, you'll also get HW-Counters statistics. These being vendor specific, please report to their documentation.
```
Did you mean `refer` instead of `report`?

Saruspete replied:
Indeed, it was a bad translation from french, where we say "reportez vous à"
thanks

vlvkobal commented on collectors/proc.plugin/sys_class_infiniband.c:
```diff
@@ -14,39 +15,37 @@
+    GEN(symbol_error, errors, "Link minor errors", 1, __VA_ARGS__)                                                     \
+    GEN(port_rcv_remote_physical_errors, errors, "Pkts rcvd with EBP", 1, __VA_ARGS__)                                 \
+    GEN(port_rcv_switch_relay_errors, errors, "Pkts rcvd discarded by switch", 1, __VA_ARGS__)                         \
+    GEN(port_xmit_constraint_errors, errors, "Pkts sent discarded by switch", 1, __VA_ARGS__)
```
Personally, I'm not fond of everything that our `.clang-format` does. I prefer to make columns aligned, as you did before.

Saruspete replied:
to be honest, when I saw the changes done by clang-format while I didn't commit before, I was so pissed... 
Just fixed it.

vlvkobal commented on collectors/proc.plugin/sys_class_infiniband.c:
```diff
@@ -558,31 +544,19 @@ int do_sys_class_infiniband(int update_every, usec_t dt) {
-                );
+                    "Infiniband", port->chart_id_packets, NULL, port->chart_family, "ib.packets", "Packets Statistics",
+                    "packets/s", PLUGIN_PROC_NAME, PLUGIN_PROC_MODULE_INFINIBAND_NAME, port->priority + 2, update_every,
+                    RRDSET_TYPE_AREA);
```
As for my taste, bin-pack isn't the best way of handling a lot of parameters. We had a relevant discussion in #7271. I can't find a good way to handle it automatically. In a case like this one, I'd prefer something like this
```
            buffer_sprintf(
                opts->wb,
                "%s_%s%s%s%s " CALCULATED_NUMBER_FORMAT " %llu\n",
                opts->prefix,
                opts->name,
                label_pre,
                opts->labels,
                label_post,
                value,
                ((rv->last_updated) ? rv->last_updated : opts->now) * 1000ULL);
```

Saruspete replied:
That's also a clang-format change :( It was like it's currently in `net_device.c` as it was my basis.
https://github.com/netdata/netdata/pull/9091/commits/3b0147825a56d82f3feae05a35b98edaf006d791#diff-b016cb1ed3cfc83e793092f14536ddb1L527-L539

Saruspete commented:
Thanks, @thiagoftsm @vlvkobal !

You can create fake hardware with the script & config I put in the "test plan" part of the PR.
